libs/ledger-live-common/src/families/cosmos/js-synchronisation.ts

```diff
@@ -54,130 +54,110 @@ const txToOps = (info: AccountShapeInfo, accountId: string, txs: CosmosTx[]): Op
 
     const correspondingMessages = messages.filter(m => m.type === mainMessage.type);
 
-    if (correspondingMessages.length === 0) {
-      continue;
-    }
-
-    // TODO: This mechanism should be removed
-    const attributes: { [id: string]: any } = {};
-    mainMessage.attributes.forEach(item => (attributes[item.key] = item.value));
-
-    // https://docs.cosmos.network/v0.42/modules/staking/07_events.html
     switch (mainMessage.type) {
       case "transfer":
-        if (attributes.sender && attributes.recipient && attributes.amount) {
-          op.senders.push(attributes.sender);
-          op.recipients.push(attributes.recipient);
-
-          if (attributes.amount.indexOf(currency.units[1].code) != -1) {
-            op.value = op.value.plus(attributes.amount.replace(currency.units[1].code, ""));
-          }
-
-          if (!op.type && attributes.sender === address) {
-            op.type = "OUT";
-            op.value = op.value.plus(fees);
-          } else if (!op.type && attributes.recipient === address) {
-            op.type = "IN";
+        // TODO: handle IBC transfers here
+        for (const message of correspondingMessages) {
+          const amount = message.attributes.find(attr => attr.key === "amount")?.value;
+          const sender = message.attributes.find(attr => attr.key === "sender")?.value;
+          const recipient = message.attributes.find(attr => attr.key === "recipient")?.value;
+          if (amount && sender && recipient && amount.indexOf(currency.units[1].code) != -1) {
+            if (op.senders.indexOf(sender) === -1) op.senders.push(sender);
```

Comment:
Why do we need to do this check ? I don't understand in what case we fill the operation and then ignore other senders, same for other similar checks

Reply:
if (op.senders.indexOf(sender) === -1) op.senders.push(sender);
-> we can make sure that there is no duplicated sender in the op.senders array.
if (op.recipients.indexOf(recipient) === -1) op.recipients.push(recipient);
-> we can make sure that there is no duplicated sender in the op.recipients array.

In the case one transaction contains multiple "transfer" events, e.g. a same sender sends funds to multiple different recipients, we don't want to display the same sender address multiple time in UI

Reply:
Why do we need to make sure there aren't duplicated sender ? let's check if it happens. We have to remove useless code